torch/csrc/api/include/torch/data/datasets/chunk.h

```diff
@@ -240,6 +249,133 @@ class BatchDataBuffer {
+  /// training, this is different to chunk_count as each loader needs to load
+  /// only a subset of chunks.
+  size_t local_chunk_count() {
+    return local_chunk_count_;
+  }
+
+ protected:
+  size_t epoch_;
+  size_t chunk_count_;
+  size_t num_replicas_;
+  size_t rank_;
+  size_t local_chunk_count_;
+};
+
+/// Select chunks randomly. The chunk order shuffled at each `reset()` call.
+class RandomChunkSelector : public ChunkSelector {
+ public:
+  RandomChunkSelector(
+      size_t chunk_count,
+      size_t num_replicas = 1,
+      size_t rank = 0)
+      : ChunkSelector(chunk_count, num_replicas, rank) {
+    size_t index_count =
+        num_replicas_ == 1 ? chunk_count_ : local_chunk_count_ * num_replicas_;
+    all_indices_.resize(index_count);
+    std::iota(std::begin(all_indices_), std::end(all_indices_), 0);
+    if (num_replicas_ > 1)
+    {
+      for (size_t i = chunk_count; i < index_count; ++i) {
+        all_indices_[i] =
+            i % chunk_count_; // we are adding some more chunks to make all
+                              // replicas to have the same number of chunks.
+      }    
+    }
+  }
+
+  optional<size_t> next() { 
+    AT_CHECK(
+        !chunk_indices_.empty(),
+        "reset() needs to be called before calling next().");
+    size_t idx = current_index_.fetch_add(1, std::memory_order_relaxed);
+    if (idx < chunk_indices_.size()) {
+      return chunk_indices_[idx];
+    } else {
+      return nullopt;
+    }
+  }
+
+  void reset() override {
+    std::minstd_rand rand(epoch_);
+    std::shuffle(all_indices_.begin(), all_indices_.end(), rand);
+    chunk_indices_.clear();
+    chunk_indices_.reserve(local_chunk_count_);
+    auto begin_index_iter = all_indices_.begin() + rank_ * local_chunk_count_;
+    auto end_index_iter = begin_index_iter + local_chunk_count_;
+    std::copy(begin_index_iter, end_index_iter, back_inserter(chunk_indices_));
+    current_index_ = 0;
+  }
+
+ private:
+  std::vector<size_t> all_indices_;
+  std::vector<size_t> chunk_indices_;
+  std::atomic<size_t> current_index_{0};
+};
+
+/// Select chunks sequentially. 
+class SequentialChunkSelector : public ChunkSelector {
+ public:
+  SequentialChunkSelector(
+      size_t chunk_count,
+      size_t num_replicas = 1,
+      size_t rank = 0)
+      : ChunkSelector(chunk_count, num_replicas, rank) {
+    begin_index_ = rank_ * local_chunk_count_;
+    end_index_ = begin_index_ + local_chunk_count_;
+    chunk_index_ = begin_index_;
+  }
+
+  optional<size_t> next() {
+    size_t idx = chunk_index_.fetch_add(1, std::memory_order_relaxed);
+    if (idx < end_index_) {
+      return idx % chunk_count_;
+    } else {
+      return nullopt;/// Select chunks randomly. The chunk order shuffled at each `reset()` call
+    }
+  }
+
+  void reset() override {
+    chunk_index_ = begin_index_;
+  }
+
+  private:
+  size_t begin_index_;
+  size_t end_index_;
+  std::atomic<size_t> chunk_index_;
```

Comment:
I don't know whether it's worth to have another non-atomic version for non-distributed system case. Currently, in both scenarios, every 'next()' call goes through the atomic chunk_index_ variable. The concurrent restriction is not necessary in the non-distributed system and would have a perf hit on the atomic lock.

Reply:
If it is a regular sampler, this could be a hit, but since this is only for chunk selection and it needs to be always thread_safe irrespective of distributed or not. ChunkSelector is used by the preloader threads in the chunkdataset.

Reply:
D'oh! yeah, you are right.